File: scripts/verificar-excel.py
"""
Segunda mitad del control del Excel.

Toma el libro que escribió `verificar-excel.ts`, lo hace recalcular por
LibreOffice y compara cada celda contra lo que calculó el motor. Además
controla que las celdas de cálculo sean fórmulas de verdad y no números
pegados: si alguien exporta un valor donde debería haber una fórmula, falla.

Uso: python3 scripts/verificar-excel.py /tmp/modelo.xlsx
"""
import json
import subprocess
import sys
import warnings
from pathlib import Path

import openpyxl

warnings.filterwarnings("ignore")

TOLERANCIA_RELATIVA = 1e-6
TOLERANCIA_ABSOLUTA = 0.01


def recalcular(origen: Path) -> Path:
    """LibreOffice abre el libro, recalcula y lo vuelve a guardar con los valores."""
    destino = origen.parent / "recalculado"
    destino.mkdir(exist_ok=True)
    subprocess.run(
        ["soffice", "--headless", "--calc", "--convert-to", "xlsx", "--outdir", str(destino), str(origen)],
        check=True, capture_output=True, timeout=600,
    )
    return destino / origen.name


def celdas_de(hoja, concepto: str):
    """Devuelve (fila de fórmulas, fila de valores) para un concepto de la columna A."""
    for fila in hoja.iter_rows(min_row=1, max_row=hoja.max_row, min_col=1, max_col=1):
        if fila[0].value == concepto:
            return fila[0].row
    return None


def main() -> int:
    ruta = Path(sys.argv[1] if len(sys.argv) > 1 else "/tmp/modelo.xlsx")
    esperado = json.loads(Path(str(ruta).replace(".xlsx", "") + ".esperado.json").read_text("utf-8"))
    anios = esperado["anios"]

    con_formulas = openpyxl.load_workbook(ruta)
    calculado = openpyxl.load_workbook(recalcular(ruta), data_only=True)

    fallas: list[str] = []
    celdas_controladas = 0
    formulas_controladas = 0

    for entrada in esperado["filas"]:
        hoja_nombre = entrada["hoja"]
        if hoja_nombre not in con_formulas.sheetnames:
            fallas.append(f"falta la hoja {hoja_nombre}")
            continue
        hf = con_formulas[hoja_nombre]
        hv = calculado[hoja_nombre]
        fila = celdas_de(hf, entrada["concepto"])
        if fila is None:
            fallas.append(f"{hoja_nombre}: no está la fila «{entrada['concepto']}»")
            continue

        for j, esperado_valor in enumerate(entrada["valores"]):
            col = 2 + j  # columna B en adelante
            formula = hf.cell(fila, col).value
            obtenido = hv.cell(fila, col).value

            if not (isinstance(formula, str) and formula.startswith("=")):
                fallas.append(
                    f"{hoja_nombre} · {entrada['concepto']} · {anios[j] if j < len(anios) else j}: "
                    f"no es una fórmula (es {formula!r})"
                )
                continue
            formulas_controladas += 1

            if esperado_valor is None:
                continue
            if obtenido is None or isinstance(obtenido, str):
                obtenido = 0.0
            celdas_controladas += 1
            escala = max(abs(esperado_valor), abs(obtenido), 1.0)
            if abs(obtenido - esperado_valor) > max(TOLERANCIA_ABSOLUTA, escala * TOLERANCIA_RELATIVA):
                fallas.append(
                    f"{hoja_nombre} · {entrada['concepto']} · "
                    f"{anios[j] if j < len(anios) else j}: planilla {obtenido:,.4f} "
                    f"vs motor {esperado_valor:,.4f}"
                )

    print(f"\nCeldas comparadas: {celdas_controladas}  ·  fórmulas verificadas: {formulas_controladas}")
    if fallas:
        print(f"\n{len(fallas)} diferencia(s):")
        for f in fallas[:40]:
            print("  FALLA", f)
        if len(fallas) > 40:
            print(f"  … y {len(fallas) - 40} más")
        return 1
    print("El Excel reproduce el modelo en todas las celdas controladas.\n")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/verificar-excel.ts
/**
 * Control del Excel exportado.
 *
 * Genera el libro, lo deja en un archivo y escribe al lado un JSON con lo que
 * calculó el motor para cada fila y cada año. Después `verificar-excel.py`
 * recalcula el libro con LibreOffice y compara celda por celda: si una fórmula
 * no reproduce el motor, el chequeo falla y se ve exactamente en qué hoja y en
 * qué año.
 *
 * Corre con: npm run check:excel
 */
import { writeFileSync } from "node:fs";
import writeXlsxFile from "write-excel-file/node";
import { escenarioBase } from "../src/lib/model/defaults";
import { calcular, kpis, tir } from "../src/lib/model/engine";
import { UNIDADES, NOMBRE_UNIDAD, Unidad } from "../src/lib/model/types";
import { construirHojas, nombreHoja } from "../src/lib/excel";

const salida = process.argv[2] ?? "/tmp/modelo.xlsx";

/**
 * Con `--con-obras` la inversión de cada unidad se carga como lista de obras en
 * lugar de un importe por año. El resultado tiene que ser exactamente el mismo:
 * si cambia algo, es que el SUMIF de la planilla o el motor no coinciden.
 */
const conObras = process.argv.includes("--con-obras");
const esc = escenarioBase();
if (conObras) {
  UNIDADES.forEach((u: Unidad) => {
    const unidad = esc.unidades[u];
    unidad.obras = unidad.capexAnual
      .map((montoMM, i) => ({ montoMM, anio: esc.base.anioBase + i }))
      .filter((x) => x.montoMM > 0)
      .map((x) => ({ id: `${u}-${x.anio}`, nombre: `Obra ${x.anio}`, ...x }));
  });
}
const c = calcular(esc);
const k = kpis(esc, c);
const hojas = construirHojas(esc, c, k);

/** Lo que debería dar cada fila del libro, según el motor. */
interface Esperado {
  hoja: string;
  concepto: string;
  valores: (number | null)[];
}

const esperado: Esperado[] = [];
const fila = (hoja: string, concepto: string, valores: (number | null)[]) =>
  esperado.push({ hoja, concepto, valores });

UNIDADES.forEach((u: Unidad) => {
  const h = nombreHoja(NOMBRE_UNIDAD[u]);
  const r = c.porUnidad[u];
  fila(h, "Toneladas potenciales sin restricción de capacidad", r.toneladasTeoricas);
  fila(h, "Toneladas efectivamente operadas", r.toneladasEfectivas);
  fila(h, "Proporción de la demanda atendida", r.factorUtilizacion);
  fila(h, "Tarifa de uso de muelle (USD/tn)", r.tarifaMuelle);
  fila(h, "Tarifa de carga y descarga (USD/tn)", r.tarifaEstibaje);
  fila(h, "Tarifa de manipuleo (USD/tn)", r.tarifaManipuleo);
  fila(h, "Tarifa de almacenaje (USD/tn)", r.tarifaAlmacenaje);
  fila(h, "Tarifa de calada y otros derechos (USD/tn)", r.tarifaCalada);
  fila(h, "TARIFA TOTAL POR TONELADA (USD/tn)", r.tarifaTotal);
  fila(h, "Facturación por uso de muelle", r.ingresosMuelle);
  fila(h, "Facturación por carga y descarga", r.ingresosEstibaje);
  fila(h, "Facturación por manipuleo", r.ingresosManipuleo);
  fila(h, "Facturación por almacenaje", r.ingresosAlmacenaje);
  fila(h, "Facturación por calada y otros derechos", r.ingresosCalada);
  fila(h, "Facturación por otros conceptos", r.ingresosOtros);
  fila(h, "FACTURACIÓN DE LA UNIDAD", r.ingresosBrutos);
  fila(h, "Costo fijo directo", r.opexFijo);
  fila(h, "Costo variable unitario (USD/tn)", r.opexVariableUnitario);
  fila(h, "Costo variable total", r.opexVariable);
  fila(h, "Costo directo total", r.opexDirecto);
  fila(h, "Costos compartidos asignados", r.opexComun);
  fila(h, "Costo operativo total", r.opexTotal);
  fila(h, "Derecho de uso portuario", r.canonTotal);
  fila(h, "GANANCIA OPERATIVA (EBITDA)", r.ebitda);
  fila(h, "Inversión directa", r.capexDirecto);
  fila(h, "Obras compartidas asignadas", r.capexComun);
  fila(h, "INVERSIÓN TOTAL DEL AÑO", r.capexTotal);
  fila(h, "Base depreciable acumulada", r.baseDepreciable);
  fila(h, "Depreciación del año", r.depreciacion);
  fila(h, "RESULTADO ANTES DE INTERESES E IMPUESTOS (EBIT)", r.ebit);
  fila(h, "[Informativo] Impuesto como sociedad independiente", r.impuestoStandalone);
  fila(h, "[Informativo] Resultado después de ese impuesto", r.nopatStandalone);
  fila(h, "FLUJO DE CAJA DE LA UNIDAD (EVALUACIÓN INDEPENDIENTE)", r.fcffStandalone);
  fila(h, "Flujo acumulado", r.fcffAcumulado);
  fila(h, "Recaladas por año", r.recaladas);
  fila(h, "Ocupación del muelle", r.ocupacionMuelle);
  fila(h, "RENDIMIENTO DE LA UNIDAD EN FORMA INDEPENDIENTE (TIR)", [tir(r.fcffStandalone)]);
});

const F = "Flujo consolidado";
fila(F, "TONELADAS TOTALES", c.toneladasTotales);
fila(F, "FACTURACIÓN TOTAL", c.ingresosBrutos);
fila(F, "Costos operativos (OPEX)", c.opexTotal);
fila(F, "Derecho de uso portuario", c.canonTotal);
fila(F, "GANANCIA OPERATIVA (EBITDA)", c.ebitda);
fila(F, "Depreciación de la inversión", c.depreciacion);
fila(F, "RESULTADO ANTES DE INTERESES E IMPUESTOS (EBIT)", c.ebit);
fila(F, "INVERSIÓN TOTAL DEL AÑO (CAPEX)", c.capexTotal);
fila(F, "Impuesto al cheque pagado (IDyCB)", c.idycbPagado);
fila(F, "Impuesto a las Ganancias determinado", c.impuestoDeterminado);
fila(F, "[Informativo] Ahorro por exención de Ingresos Brutos", c.memoAhorroIIBB);
fila(F, "[Informativo] Ahorro por exención municipal", c.memoAhorroMunicipal);
fila(F, "Impuesto al cheque computado a cuenta", c.ahorroDebCred);
fila(F, "Impuesto al cheque recuperado", c.idycbRecuperado);
fila(F, "Impuesto a las Ganancias a pagar", c.impuestoNeto);
fila(F, "Tasa municipal de Timbúes (DREI)", c.drei);
fila(F, "Tasa de edificación y movimiento de tierra", c.tasaEdificacion);
fila(F, "[Informativo] IVA de las inversiones (CERTIVA)", c.memoIVAInversiones);
fila(F, "RESULTADO OPERATIVO DESPUÉS DE IMPUESTOS (NOPAT)", c.nopat);
fila(F, "FLUJO DE CAJA LIBRE DEL PROYECTO (FCFF)", c.fcff);
fila(F, "Desembolso del préstamo", c.deudaDesembolso);
fila(F, "Intereses del préstamo", c.deudaIntereses);
fila(F, "Amortización del capital", c.deudaAmortizacion);
fila(F, "Saldo de la deuda", c.deudaSaldo);
fila(F, "Escudo fiscal de los intereses", c.escudoFiscal);
fila(F, "FLUJO PARA LOS ACCIONISTAS (FCFE)", c.fcfe);
fila(F, "Servicio de la deuda (capital + intereses)", c.servicioDeuda);
fila(F, "Flujo del proyecto acumulado", c.fcffAcumulado);
fila(F, "Flujo para los accionistas acumulado", c.fcfeAcumulado);
fila(F, "OCUPACIÓN DEL MUELLE", c.ocupacionMuelle);
fila(F, "RENDIMIENTO DEL PROYECTO (TIR)", [k.tirProyecto]);
UNIDADES.forEach((u: Unidad) => {
  const corto = u === "AGRO" ? "agrograneles" : u === "FERT" ? "fertilizantes y líquidos" : "cargas generales";
  fila(F, `Toneladas de ${corto}`, c.porUnidad[u].toneladasEfectivas);
  fila(F, `Facturación de ${corto}`, c.porUnidad[u].ingresosBrutos);
  fila(F, `Inversión en ${corto}`, c.porUnidad[u].capexTotal);
  fila(F, `   Flujo que corresponde a ${corto}`, c.fcffPorUnidad[u]);
});

fila("Resumen", "Rendimiento del proyecto (TIR)", [k.tirProyecto]);
fila("Resumen", "Inversión total (CAPEX)", [k.capexTotal]);
fila("Resumen", "Resultado operativo acumulado (EBITDA)", [k.ebitdaAcumulado]);
fila("Resumen", "Facturación acumulada", [k.ingresosAcumulados]);
fila("Resumen", "Margen operativo", [k.margenEbitda]);
fila("Resumen", "Toneladas del año pico", [k.toneladasMaximas]);
fila("Resumen", "Ocupación máxima del muelle", [k.ocupacionMaxima]);
fila("Resumen", "Inversión por tonelada instalada", [k.capexPorToneladaInstalada]);
UNIDADES.forEach((u: Unidad) =>
  fila("Resumen", NOMBRE_UNIDAD[u], [tir(c.porUnidad[u].fcffStandalone)]));

writeFileSync(salida.replace(/\.xlsx$/, "") + ".esperado.json",
  JSON.stringify({ anios: c.anios, filas: esperado }, null, 1), "utf-8");

const libro = writeXlsxFile(hojas as never) as unknown as { toFile: (p: string) => Promise<void> };
libro.toFile(salida)
  .then(() => console.log(`Libro escrito en ${salida} (${esperado.length} filas a controlar).`))
  .catch((e: unknown) => { console.error("FALLO al escribir:", e); process.exit(1); });
